feat: restructure repo & add TypeScript types & add ESM exports & add json schema

favna opened the pull request:
This PR restructures the repo a fair bit, moving a bunch of files into a dedicated `src` directory. Other notable changes are:

- `presenter.js` is moved to a `bin` directory
- TypeScript types are added in `typings` directory
- JSON schema is added in `assets` directory
  - The initial file was autogenerated, however I have since gone through it all manually and resolving issues as requested.
  - Editors such as VSCode can read the file and give code completion based on the schema when given like so:
  ```json
  {
      "$schema": "https://raw.githubusercontent.com/joellord/cli-slides/master/assets/json-schema.json"
  }
  ```
   (or `https://raw.githubusercontent.com/Favna/cli-slides/feat/add-json-schema/assets/json-schema.json` if you want to test it)
- ESM exports are added in `esm` directory
- package.json has been changed a fair bit, not removing any information from but only adding.
  - In particular the `files` array to limit which files are to be put in the tarball that is uploaded to npm
  - Also in particular all the export bindings (`main`, `module`, `types`, `exports` and `bin` keys)
- Added a `LICENSE.md` file
- Improved `README.md` a lot

favna committed:
feat: add json schema

It's probably not perfect but it helps anyway.

favna committed:
docs: update readme to reference json schema

favna committed:
fix(schema): update schema with manual changes

favna committed:
fix(schema): one more change

joellord commented:
This look good and I really like the idea of having a $schema for the JSON slides. However, the way it is built right now, it's expecting all the slide types in the format specified in the proposed schema. 
I am not familiar with json-schema, but my understanding is that it might need a schema for each slide type. Then "items" would "contain" "anyOf" the new types.
I will look into this soon-ish. Feel free to comment if I have a bad understanding of how it works.

favna committed:
feat: fix schema, improve readme, add license file

favna commented:
@joellord it took me quite a while to get around it but it was worth the wait I'd say.

The schema is completely fixed now. As soon as a `type` is provided in the slides, all the other properties, descriptions and other schema keys are resolved like magic. 

I also took the liberty of rewriting big chunks of the README, in particular a bunch of Markdown formatting (lists, syntax highlighting, etc). I also improved the "header" part a bit to be more catchy and for my own slides I wrote a simple JavaScript file to start the presentation so I can use `node /path/to/file.mjs`, figured I'd add a section about that as well.

Then I also added a LICENSE.md for the ISC license that's on this project and put the year 2019 in there as that's when you first committed to this repo.

And lastly, I moved a bunch of files around to improve the final bundle that'll be on npmjs. Dropped the actual "source" files in `src`, created a directory `bin` which has `presenter.js`, created a directory `esm` which has the ESM compatible exports for the `presentation.js` file for easy importing it in ESM, and created a directory `typings` which has TypeScript typings.

I have also updated the original PR description and PR title to reflect all of this.

favna committed:
docs(readme): fix license badge

Apparently GitHub doesn't recognize the ICS license

favna committed:
feat: restructure repo

- Add TypeScript typings
- Add ESM exports
- Overall improve package.json
- Add files array to package.json

favna committed:
chore: use proper name for ESM export

kyranet commented on typings/presentation.d.ts:
```diff
@@ -0,0 +1,13 @@
+    public start(startSheet?: string | number): void;
+  }
+  
+  
```
```suggestion
```

lveneris commented on README.md:
```diff
@@ -1,56 +1,164 @@
+
+### Starting through JavaScript
+
+If you would prefer to start your presentation through JavaScript, for example if you want to be able to call `node start-slides.js` then you can create a script like this:
```
```suggestion
If you would prefer to start your presentation through JavaScript, for example if you want to be able to call `node start-slides.js`, then you can create a script like this:
```

lveneris commented on README.md:
```diff
@@ -1,56 +1,164 @@
+
+## About
+
+CLI Slides is a framework to build slide decks that will run in the terminal. This application will use a JSON file to generate a full slide deck. You can use many of the different templates for the slides or, if you feel adventurous, you can create your own.
```
```suggestion
CLI Slides is a framework to build slide decks that will run in the terminal. This application will use a JSON file to generate a full slide deck. You can use many of the different templates for the slides - or, if you feel adventurous, you can even create your own.
```

lveneris commented on README.md:
```diff
@@ -1,56 +1,164 @@
+
+CLI Slides is a framework to build slide decks that will run in the terminal. This application will use a JSON file to generate a full slide deck. You can use many of the different templates for the slides or, if you feel adventurous, you can create your own.
 
 Once you start the presentation tool, it will support hot reloading of you slide deck. This means you can open your presentation, change the JSON file and you should immediately see the changes in your terminal.
```
```suggestion
Once you start the presentation tool, it comes with hot reloading for your slide deck, so you can change the JSON file and preview your changes on the fly.
```

lveneris commented on README.md:
```diff
@@ -1,56 +1,164 @@
+
+## Usage
+
+After infalling `cli-slides` you can start your slidedeck by passing it the slidedeck JSON file as first argument:
```
```suggestion
After installing `cli-slides` you can start your slidedeck by passing it the slidedeck JSON file as the first argument, like so:
```

lveneris commented on README.md:
````diff
@@ -1,56 +1,164 @@
 cli-slides path/to/presentation.json
 ```
 
+Optionally you can also provide a second argument to directly jump to the slide with that index (1-based):
````
```suggestion
Optionally, you can also provide a second argument to directly jump to the slide with the specified index (starting from 1):
```

lveneris commented on README.md:
```diff
@@ -1,56 +1,164 @@
+
+### JSON Schema
 
+A Json schema is available in [./assets/json-schema.json](/assets/json-schema.json). You can use it in your JSON like so:
```
```suggestion
A JSON schema is available at [`assets/json-schema.json`](assets/json-schema.json). You can use it in your JSON like so:
```

lveneris commented on README.md:
```diff
@@ -1,56 +1,164 @@
-You can specify what to display at the top of the slides, in the center by setting the frame.top.center property. CLI-slides will try to find a matching property in the deck definition or will display the text as a specified.
 
-You can also customize the bottom part. The bottom section takes up to two lines. The frame.bottom.left and frame.bottom.center properties both take a string (for a single line) or an array of strings (if you need two lines). The strings can either be the property to display in the deck definition file or some text to be displayed.
+You can specify what to display at the top of the slides, in the center by setting the `frame.top.center` property. CLI-slides will try to find a matching property in the deck definition or will display the text as a specified.
```
```suggestion
You can specify content to display centered at the top with the `frame.top.center` property. First, `cli-slides` will attempt to find a property matching the value you set, but if one doesn't exist, it will display the specified text.
```

lveneris commented on README.md:
```diff
@@ -62,16 +170,18 @@ A slide template to display code. Code be be displayed as a whole, line by line,
-*multistep* (optional): boolean.
-*multistep* (optional): "line" or "highlight". Will display one line at a time or one highlighted item at a time.
+- `title`: The title to be displayed on this slide.
+- `code`: The code snippet to be displayed, will be enclosed in a box. Use an array to display multiple lines of code.
```
```suggestion
- `code`: The code snippet to be displayed inside of a code block. Use an array to display multiple lines of code.
```

lveneris commented on README.md:
```diff
@@ -62,16 +170,18 @@ A slide template to display code. Code be be displayed as a whole, line by line,
+- `code`: The code snippet to be displayed, will be enclosed in a box. Use an array to display multiple lines of code.
+- `text`: Description of the code snippet, displayed over the box.
+- `notes` (optional): Adds some notes at the bottom of the slide.
+- `multistep` (optional): boolean.
```
```suggestion
- `multistep` (optional): A boolean to toggle multistep mode.
```

lveneris commented on README.md:
```diff
@@ -62,16 +170,18 @@ A slide template to display code. Code be be displayed as a whole, line by line,
+- `text`: Description of the code snippet, displayed over the box.
+- `notes` (optional): Adds some notes at the bottom of the slide.
+- `multistep` (optional): boolean.
+- `multistepType` (optional): "line" or "highlight". Will display one line at a time or one highlighted item at a time.
```
```suggestion
- `multistepType` (optional): This can take two values; "line", or "highlight". "line" displays one line at a time, "highlight" displays one highlighted item at a time.
```

lveneris commented on README.md:
```diff
@@ -62,16 +170,18 @@ A slide template to display code. Code be be displayed as a whole, line by line,
+#### Diagram
 
-### Diagram
 Used to display some text on the left side and some ASCII art on the right side. You're on your own to draw that diagram though.
```
```suggestion
Used to display some text on the left side and some ASCII art on the right side. Drawing the diagram is your responsibility, though - there are none provided with `cli-slides`.
```

lveneris commented on README.md:
```diff
@@ -83,146 +193,152 @@ Used to display some text on the left side and some ASCII art on the right side.
-*title*: The title to be displayed on the left side of the slide.
-*diagram*: An array of strings to be displayed on the right side
+- `title`: The title to be displayed on the left side of the slide.
+- `diagram`: An array of strings to be displayed on the right side.
```
```suggestion
- `diagram`: An array of strings to be displayed as lines of ASCII art on the right of the slide.
```

lveneris commented on README.md:
```diff
@@ -83,146 +193,152 @@ Used to display some text on the left side and some ASCII art on the right side.
+#### List
 
-### List
 A list of items. Items can be displayed at once or one at a time.
```
```suggestion
A list of items. Items can either be displayed one at a time, or all at once.
```

lveneris commented on README.md:
```diff
@@ -83,146 +193,152 @@ Used to display some text on the left side and some ASCII art on the right side.
-*title*: The title to be displayed on this slide. 
-*text*: Some text to be displayed. Text will be centered and wraps automatically. Supports multiple lines with the \n character
+- `title`: The title to be displayed on this slide.
+- `text`: Some text to be displayed. Text will be centred and wraps automatically. Supports multiple lines with the \n character
```
```suggestion
- `text`: Some text to be displayed. Text will be centred and wraps automatically. Supports multiple lines with a newline escape (`\n`).
```

lveneris commented on README.md:
```diff
@@ -83,146 +193,152 @@ Used to display some text on the left side and some ASCII art on the right side.
+- `character`: The preset ASCII character to display
+  - `over-cubbie` Someone looking over a wall
+  - `me`: A man smiling
+  - `me-oh-no`: Same as me with closed eyes and open mouth
```
```suggestion
  - `me-oh-no`: The same as `me`, but with closed eyes and open mouth
```

lveneris commented on README.md:
```diff
@@ -83,146 +193,152 @@ Used to display some text on the left side and some ASCII art on the right side.
+  - `me-oh-no`: Same as me with closed eyes and open mouth
+  - `silly-face`: A face with a confused look
+  - `scream`: The scream emoji converted to ASCII art
+- `text`: Text to be displayed in a bubble
```
```suggestion
- `text`: Text to be displayed next to the character in a speech bubble
```

lveneris commented on README.md:
```diff
@@ -83,146 +193,152 @@ Used to display some text on the left side and some ASCII art on the right side.
+
+#### Terminal
+
 The famous terminal inside the terminal!
```
```suggestion
The famous terminal-in-a-terminal! ~~(obligatory inception joke goes here)~~
```

lveneris commented on README.md:
```diff
@@ -83,146 +193,152 @@ Used to display some text on the left side and some ASCII art on the right side.
+#### TitleOnly
 
-### TitleOnly
 Just a title slide. Can be used for section separator.
```
```suggestion
Just a title slide, typically used as a section separator.
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+        "multistep": {
+          "type": "boolean",
+          "default": false,
+          "description": "Whether this slide should show it's content in multiple steps, advancing a step every time the enter key is pressed."
```
```suggestion
          "description": "Whether this slide should show its content in multiple steps, advancing a step every time the enter key is pressed."
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+  "required": ["title", "presenter", "date", "conference", "location", "company", "twitter", "slides"],
+  "definitions": {
+    "codeSlide": {
+      "description": "A slide template to display code. Code be be displayed as a whole, line by line, or one highlighted section at a time.",
```
```suggestion
      "description": "A slide template to display code. Code can be displayed as a whole, line by line, or one highlighted section at a time.",
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+          "description": "Description of the code snippet, displayed over the box"
+        },
+        "code": {
+          "description": "The code snippet to be displayed, will be enclosed in a box. Use an array to display multiple lines of code.",
```
```suggestion
          "description": "The code snippet to be displayed in a code block. Use an array to display multiple lines of code.",
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+      "required": ["title", "text", "code"]
+    },
+    "diagramSlide": {
+      "description": "Used to display some text on the left side and some ASCII art on the right side. You're on your own to draw that diagram though.",
```
```suggestion
      "description": "Used to display some text on the left side and some ASCII art on the right side. The ASCII art is down to you though, there are none provided by `cli-slides`.",
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+        "text": {
+          "type": "string",
+          "default": "",
+          "description": "Some text to be displayed. Text will be centered and wraps automatically. Supports multiple lines with the \n character"
```
```suggestion
          "description": "Some text to be displayed. Text will be centered and wraps automatically. Supports multiple lines with a newline escape (\n).
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+          "type": "string",
+          "enum": ["over-cubbie", "me", "me-oh-no", "silly-face", "scream"],
+          "title": "The preset ASCII character to display",
+          "description": "- over-cubbie: Someone looking over a wall\n- me: A man smiling\n- me-oh-no: Same as me, with closed eyes and open mouth\n- silly-face: A face with confused look\n- scream: The scream emoji converted to ASCII art"
```
```suggestion
          "description": "- over-cubbie: Someone looking over a wall\n- me: A man smiling\n- me-oh-no: Same as me, with closed eyes and an open mouth\n- silly-face: A face with confused look\n- scream: The scream emoji converted to ASCII art"
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+      "required": ["left", "right"]
+    },
+    "terminalSlide": {
+      "title": "The famous terminal inside the terminal!",
```
```suggestion
      "title": "The famous terminal-in-a-terminal!",
```

lveneris commented on assets/json-schema.json:
```diff
@@ -0,0 +1,352 @@
+      "required": ["title"]
+    },
+    "titleOnlySlide": {
+      "description": "Just a title slide. Can be used for section separator.",
```
```suggestion
      "description": "Just a title slide, typically used as a section separator.",
```

Nytelife26 reviewed (changes requested):
Just grammar nitpicks, as per usual.

favna committed:
Apply suggestions from code review

Thanks for the grammar checking, friends <3

Co-authored-by: Tyler J Russell <xtylerjrx@gmail.com>

favna committed:
Update typings/presentation.d.ts

Co-authored-by: Antonio Román <kyradiscord@gmail.com>

favna commented:
This is why having friends and talking about stuff I'm working on is great. Thanks for the grammar/format checking @Nytelife26 and @kyranet ❤️

joellord commented:
Thanks for the contribution!

favna commented:
@joellord Just 1 request - could you also make a release to NPM, so the typings and ESM export are available there? I have a presentation planned soon-ish for my colleagues and what the slides I have prepped so far are in fact using this lib 😉 

(p.s. some of the info is what I learned at iJS 😛)

joellord commented:
Done! I had to adjust the version numbers so the official release is 1.2.0 now. 

Let me know on Twitter (https://twitter.com/joel__lord) how the presentation goes! 🎉
